README.md
# Vagrant Cosmic Provider

This is a fork of the [CloudStack Vagrant provider](https://github.com/MissionCriticalCloud/vagrant-cloudstack).

This is a [Vagrant](http://www.vagrantup.com) 2.2+ plugin that adds a `cosmic`
provider to Vagrant for use with [Cosmic](https://github.com/MissionCriticalCloud/cosmic).

## Features

* SSH into the instances.
* Provision the instances with any built-in Vagrant provisioner.
* Minimal synced folder support via `rsync`/`winrm`.

## Usage

Install using standard Vagrant 2.2+ plugin installation methods. After
installing, `vagrant up` and specify the `cosmic` provider. An example is
shown below.

```shell
$ vagrant plugin install vagrant-cosmic

...

$ vagrant up --provider=cosmic

...
```

## Quick Start

After installing the plugin (instructions above), the quickest way to get
started is to actually make a Vagrantfile that looks like the following, filling in
your information where necessary.

```ruby
Vagrant.configure("2") do |config|
  config.vm.box = "${cosmic.template_name}"

  config.vm.provider :cosmic do |cosmic, override|
    cosmic.host = "cosmic.local"
    cosmic.path = "/client/api"
    cosmic.port = "8080"
    cosmic.scheme = "http"
    cosmic.api_key = "AAAAAAAAAAAAAAAAAAA"
    cosmic.secret_key = "AAAAAAAAAAAAAAAAAAA"

    cosmic.service_offering_id = "AAAAAAAAAAAAAAAAAAA"
    cosmic.disk_offering_id = "AAAAAAAAAAAAAAAAAAA"
    cosmic.network_id = "AAAAAAAAAAAAAAAAAAA"
    cosmic.zone_id = "AAAAAAAAAAAAAAAAAAA"
    cosmic.project_id = "AAAAAAAAAAAAAAAAAAA"
  end
end
```

Or with names instead of ids:

```ruby
Vagrant.configure("2") do |config|
  config.vm.box = "${cosmic.template_name}"

  config.vm.provider :cosmic do |cosmic, override|
    cosmic.host = "cosmic.local"
    cosmic.path = "/client/api"
    cosmic.port = "8080"
    cosmic.scheme = "http"
    cosmic.api_key = "AAAAAAAAAAAAAAAAAAA"
    cosmic.secret_key = "AAAAAAAAAAAAAAAAAAA"

    cosmic.service_offering_name = "THE-BESTEST"
    cosmic.disk_offering_name = "THE-LARGEST-OFFER-AVAILABLE"
    cosmic.network_name = "WOW-SUCH-FAST-OFFERING"
    cosmic.zone_name = "District-9"
    cosmic.name = "doge-is-a-hostname-now"
    # Sadly there is currently no support for the project API in fog.
    cosmic.project_id = "AAAAAAAAAAAAAAAAAAA"
  end
end
```

And then run `vagrant up --provider=cosmic`.

This will start an instance in Cosmic. And assuming your template
on Cosmic is Vagrant compatible _(vagrant user with official
vagrant pub key in authorized_keys)_ SSH and provisioning will work as
well.

## Configuration

This provider exposes quite a few provider-specific configuration options. Most of the settings
have both an id and a name setting and you can chose to use either (i.e network_id or network_name).
This gives the possibility to use the easier to remember name instead of the UUID,
this will also enable you to upgrade the different settings in your cloud without having
to update UUIDs in your Vagrantfile. If both are specified, the id parameter takes precedence.

* `name` - Hostname of the created machine
* `host` - Cosmic API host
* `path` - Cosmic API path
* `port` - Cosmic API port
* `scheme` - Cosmic API scheme _(defaults: https (thanks to the resolution order in fog))_
* `api_key` - The API key for accessing Cosmic
* `secret_key` - The secret key for accessing Cosmic
* `affinity_group_id` - An affinity group uuid or array of uuid(s) to add the instance to
* `affinity_group_name` - An affinity group name or array of name(s) to add the instance to
* `domain_id` - Domain id to launch the instance into
* `instance_ready_timeout` - The number of seconds to wait for the instance to become "ready" in Cosmic. Defaults to 120 seconds.
* `network_id` - Network uuid(s) that the instance should use
  * `network_id` is single value (e.g. `"AAAA"`) or multiple values (e.g. `["AAAA", "BBBB"]`)
* `network_name` - Network name(s) that the instance should use
  * `network_name` is single value (e.g. `"AAAA"`) or multiple values (e.g. `["AAAA", "BBBB"]`)
* `project_id` - Project uuid that the instance should belong to
* `service_offering_id`- Service offering uuid to use for the instance
* `service_offering_name`- Service offering name to use for the instance
* `template_id` - Template uuid to use for the instance
* `template_name` - Template name to use for the instance, defaults to Vagrants config.vm.box
* `zone_id` - Zone uuid to launch the instance into
* `zone_name` - Zone uuid to launch the instance into
* `keypair` - SSH keypair name, if neither'keypair' nor 'ssh_key' have been specified, a temporary keypair will be created
* `static_nat` - static nat for the virtual machine
* `pf_ip_address_id` - IP address ID for port forwarding rule
* `pf_ip_address` - IP address for port forwarding rule
* `pf_public_port` - Public Communicator port for port forwarding rule
* `pf_public_rdp_port` - Public RDP port for port forwarding rule
* `pf_public_port_randomrange` - If public port is omited, a port from this range wll be used (default `{:start=>49152, :end=>65535}`)
* `pf_private_port` - Private port for port forwarding rule (defaults to respective Communicator protocol)
* `pf_open_firewall` - Flag to enable/disable automatic open firewall rule (by Cosmic)
* `pf_trusted_networks` - Array of CIDRs or (array of) comma-separated string of CIDRs to network(s) to
  * automatically (by plugin) generate firewall rules for, ignored if `pf_open_firewall` set `true`
  * use as default for firewall rules where source CIDR is missing
* `port_forwarding_rules` - Port forwarding rules for the virtual machine
* `firewall_rules` - Firewall rules
* `display_name` - Display name for the instance
* `group` - Group for the instance
* `ssh_key` - Path to a private key to be used with ssh _(defaults to Vagrant's `config.ssh.private_key_path`)_
* `ssh_user` - User name to be used with ssh _(defaults to Vagrant's `config.ssh.username`)_
* `ssh_network_id` - The network_id to be used when loging in to the vm via ssh _(defaults to first nic)_
* `ssh_network_name` - The network_name to be used when loging in to the vm via ssh _(defaults to first nic)_
  * Use either `ssh_network_id` or `ssh_network_name`. If specified both , use `ssh_network_id`
* `vm_user` - User name to be used with winrm _(defaults to Vagrant's `config.winrm.username`)_
* `vm_password` - Password to be used with winrm. _(If the Cosmic template is "Password Enabled", leaving this unset will trigger the plugin to retrieve and use it.)_
* `private_ip_address` - private (static)ip address to be used by the virtual machine
* `expunge_on_destroy` - Flag to enable/disable expunge vm on destroy

These can be set like typical provider-specific configuration:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cosmic do |cosmic|
    cosmic.api_key = "foo"
    cosmic.secret_key = "bar"
  end
end
```

In addition to the above top-level configs, you can use the `region_config` method to specify region-specific overrides within your Vagrantfile. Note that the top-level `region` config must always be specified to choose which region you want to actually use, however. This looks like this:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cosmic do |cosmic|
    cosmic.api_key = "foo"
    cosmic.secret_key = "bar"
    cosmic.domain = "internal"

    # Simple domain config
    cosmic.domain_config "internal", :network_id => "AAAAAAAAAAAAAAAAAAA"

    # More comprehensive region config
    cosmic.domain_config "internal" do |domain|
      domain.network_id = "AAAAAAAAAAAAAAAAAAA"
      domain.service_offering_id = "AAAAAAAAAAAAAAAAAAA"
    end
  end
end
```

The domain-specific configurations will override the top-level configurations when that domain is used. They otherwise inherit the top-level configurations, as you would probably expect.

## Networks

Networking features in the form of `config.vm.network` are not supported with `vagrant-cosmic`, currently. If any of these are specified, Vagrant will emit a warning, but will otherwise boot the Cosmic machine.

### Basic networking versus Advanced networking

The plugin will determine this network type dynamically from the zone. The setting `network_type` in the Vagrant file has been deprecated, and is silently ignored.

### Basic Networking

If the network type of your zone is `basic`, you can use Security Groups and associate rules in your Vagrantfile.

If you already have Security Groups, you can associate them to your instance, with their IDs:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cosmic do |cosmic|
    cosmic.api_key = "foo"
    cosmic.secret_key = "bar"
    cosmic.security_group_ids = ['aaaa-bbbb-cccc-dddd', '1111-2222-3333-4444']
  end
end
```

or their names:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cosmic do |cosmic|
    cosmic.api_key = "foo"
    cosmic.secret_key = "bar"
    cosmic.security_group_names = ['
min_fantastiska_security_group', 'another_security_grupp']
  end
end
```

But you can also create your Security Groups in the Vagrantfile:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cosmic do |cosmic|
    cosmic.api_key = "foo"
    cosmic.secret_key = "bar"
    cosmic.security_groups = [
      {
        :name         => "Awesome_security_group",
        :description  => "Created from the Vagrantfile",
          :rules => [
        {:type => "ingress", :protocol => "TCP", :startport => 22, :endport => 22, :cidrlist => "0.0.0.0/0"},
        {:type => "ingress", :protocol => "TCP", :startport => 80, :endport => 80, :cidrlist => "0.0.0.0/0"},
        {:type => "egress",  :protocol => "TCP", :startport => 81, :endport => 82, :cidrlist => "1.2.3.4/24"},
    ]
      }
    ]
  end
end
```

### Static NAT, Firewall, Port forwarding

You can create your static nat, firewall and port forwarding rules in the Vagrantfile. You can use this rule to access virtual machine from an external machine.

The rules created in Vagrantfile are removed when the virtual machine is destroyed.

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cosmic do |cosmic|

    override.ssh.host = "X.X.X.X"

    cosmic.static_nat = [
      { :ipaddress => "A.A.A.A"}
    ]

    cosmic.port_forwarding_rules = [
      { :ipaddress => "X.X.X.X", :protocol => "tcp", :publicport => 22, :privateport  => 22, :openfirewall => false },
      { :ipaddress => "X.X.X.X", :protocol => "tcp", :publicport => 80, :privateport  => 80, :openfirewall => false }
    ]

    cosmic.firewall_rules = [
      { :ipaddress => "A.A.A.A", :cidrlist  => "1.2.3.4/24", :protocol => "icmp", :icmptype => 8, :icmpcode => 0 },
      { :ipaddress => "X.X.X.X", :cidrlist  => "1.2.3.4/24", :protocol => "tcp", :startport => 22, :endport => 22 },
      { :ipaddress => "X.X.X.X", :cidrlist  => "1.2.3.4/24", :protocol => "tcp", :startport => 80, :endport => 80 }
    ]

  end
end
```

Most values in the firewall and portforwarding rules are not mandatory, except either startport/endport or privateport/publicport

* `:ipaddress` - defaults to `pf_ip_address`
* `:protocol` - defaults to `'tcp'`
* `:publicport` - defaults to `:privateport`
* `:privateport` - defaults to `:publicport`
* `:openfirewall` - defaults to `pf_open_firewall`
* `:cidrlist` - defaults to `pf_trusted_networks`
* `:startport` - defaults to `:endport`
* `:endport` - not required by Cosmic

For only allowing Vagrant to access the box for further provisioning (SSH/WinRM), and opening the Firewall for some subnets, the following config is sufficient:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cosmic do |cosmic|
    cosmic.pf_open_firewall      = "false"
    cosmic.pf_ip_address         = X.X.X.X
    cosmic.pf_trusted_networks   = [ "1.2.3.4/24" , "11.22.33.44/32" ]
  end
end
```

Where X.X.X.X is the ip of the respective Cosmic network, this will automatically map the port of the used Communicator (SSH/Winrm) via a random public port, open the Firewall and set Vagrant to use it.

The plugin can also automatically generate firewall rules off of the portforwarding rules:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cosmic do |cosmic|

    cosmic.pf_trusted_networks   = "1.2.3.4/24,11.22.33.44/32"
    cosmic.port_forwarding_rules = [
      { :privateport  => 22, :generate_firewall => true },
      { :privateport  => 80, :generate_firewall => true }
    ]

  end
end
```

### Virtual Router versus VPC

Both Virtual Routers and VPCs are supported when using port-forwarding and firewall. This is automatically determined by the specific `pf_ip_address`.

Note that there are architectural differences in Cosmic which the configuration must adhere to.

For VPC:

* `pf_open_firewall` will be ignored as global setting and (specifically) in `port_forwarding_rules`
* for `firewall_rules` to open access for `port_forwarding_rules`, the firewall rule should allow traffic for the `:privateport` port.

For Virtual Router:

* for `firewall_rules` to open access for `port_forwarding_rules`, the firewall rule should allow traffic for the `:publicport` port.

Usage of other attributes and features work with both network types. Such as `:generate_firewall` for portforwarding rules, or `pf_trusted_networks` to automatically generate rules for the Communicator.

## Synced Folders

There is minimal support for synced folders. Upon `vagrant up`, `vagrant reload`, and `vagrant provision`, the Cosmic provider will use `rsync` (if available) to uni-directionally sync the folder to the remote machine over SSH, and use Vagrant plugin `vagrant-winrm-syncedfolders` (if available) to uni-directionally sync the folder to the remote machine over WinRM.

This is good enough for all built-in Vagrant provisioners (shell,
chef, and puppet) to work!

### User data

You can specify user data for the instance being booted.

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cosmic do |cosmic|
    # Option 1: a single string
    cosmic.user_data = "#!/bin/bash\necho 'got user data' > /tmp/user_data.log\necho"

    # Option 2: use a file
    cosmic.user_data = File.read("user_data.txt")
  end
end
```

The maximum length of user_data is around 1500 bytes with Cosmic API < 4.2 (base64 encoded user_data must be < 2048 bytes)

## Development

To work on the `vagrant-cosmic` plugin, clone this repository out, and use
[Bundler](http://gembundler.com) to get the dependencies:

```shell
bundle install
```

Once you have the dependencies, verify the unit tests pass with `rake`:

```shell
bundle exec rake
```

If the unit-tests pass, verify the plugin is functionaly good by running the functional tests with bats. Before running the tests you need to export a set of variables that are used in the tests. Look at the Rake file for the required variables, or run the following Rake command to check:

```shell
bundle exec rake functional_tests:check_environment
```

Run all functional tests by executing:

```shell
bundle exec rake functional_tests:all
```

If those pass, you're ready to start developing the plugin. You can test
the plugin without installing it into your Vagrant environment by just
creating a `Vagrantfile` in the top level of this directory (it is gitignored)
and add the following line to your `Vagrantfile`:

```ruby
Vagrant.require_plugin "vagrant-cosmic"
```

Use `bundler` to execute Vagrant:

```shell
bundle exec vagrant up --provider=cosmic
```


CHANGELOG.md
# Cosmic Vagrant plugin

This file is used to list changes made in each version of the `vagrant-cosmic` plugin.

## 0.2.0 (2020-06-11)

- Added support to provision instances in an already existing Affinity Group(s).

## 0.1.0 (2020-02-25)

- Forked plugin from [MissionCriticalCloud/vagrant-cloudstack](https://github.com/MissionCriticalCloud/vagrant-cloudstack) and renamed to `vagrant-cosmic`
- Removed Security Group functionality as it's no longer supported by the [Cosmic API](http://apidoc.mcc.schubergphilis.com)


lib/vagrant-cosmic/config.rb
require "vagrant"

module VagrantPlugins
  module Cosmic
    class Config < Vagrant.plugin("2", :config)
      INSTANCE_VAR_DEFAULT_NIL = %w(affinity_group_id
                                    affinity_group_name
                                    display_name
                                    domain_id
                                    group
                                    host
                                    keypair
                                    name
                                    network_id
                                    network_name
                                    path
                                    pf_ip_address
                                    pf_ip_address_id
                                    pf_private_port
                                    pf_public_port
                                    pf_public_rdp_port
                                    pf_trusted_networks
                                    port
                                    private_ip_address
                                    project_id
                                    service_offering_id
                                    service_offering_name
                                    ssh_key
                                    ssh_network_id
                                    ssh_network_name
                                    ssh_user
                                    template_id
                                    template_name
                                    user_data
                                    vm_password
                                    vm_user
                                    zone_id
                                    zone_name).freeze
      INSTANCE_VAR_DEFAULT_EMPTY_ARRAY = %w(static_nat port_forwarding_rules firewall_rules).freeze

      ### API settings

      # Cosmic API host.
      #
      # @return [String]
      attr_accessor :host

      # Cosmic API path.
      #
      # @return [String]
      attr_accessor :path

      # Cosmic API port.
      #
      # @return [String]
      attr_accessor :port

      # Cosmic API scheme
      #
      # @return [String]
      attr_accessor :scheme

      # The API key for accessing Cosmic.
      #
      # @return [String]
      attr_accessor :api_key

      # The secret key for accessing Cosmic.
      #
      # @return [String]
      attr_accessor :secret_key

      ### Instance settings

      # Hostname for the machine instance
      # This will be passed through to the api.
      #
      # @return [String]
      attr_accessor :name

      # Affinity group ID(s) the instance should be applied to
      #
      # @return [String]
      attr_accessor :affinity_group_id

      # Affinity group name(s) the instance should be applied to
      #
      # @return [String]
      attr_accessor :affinity_group_name

      # Disk offering uuid to use for the instance
      #
      # @return [String]
      attr_accessor :disk_offering_id

      # Disk offering name to use for the instance
      #
      # @return [String]
      attr_accessor :disk_offering_name

      # display name for the instance
      #
      # @return [String]
      attr_accessor :display_name

      # Domain id to launch the instance into.
      #
      # @return [String]
      attr_accessor :domain_id

      # flag to enable/disable expunge vm on destroy
      #
      # @return [Boolean]
      attr_accessor :expunge_on_destroy

      # group for the instance
      #
      # @return [String]
      attr_accessor :group

      # The timeout to wait for an instance to become ready.
      #
      # @return [Fixnum]
      attr_accessor :instance_ready_timeout

      # The name of the keypair to use.
      #
      # @return [String]
      attr_accessor :keypair

      # Network uuid(s) that the instance should use
      #
      # @return [String,Array]
      attr_accessor :network_id

      # Network name(s) that the instance should use
      #
      # @return [String,Array]
      attr_accessor :network_name

      # Network Type
      #
      # @return [String]
      attr_accessor :network_type

      # Private ip for the instance
      #
      # @return [String]
      attr_accessor :private_ip_address

      # Project uuid that the instance should belong to
      #
      # @return [String]
      attr_accessor :project_id

      # Service offering uuid to use for the instance
      #
      # @return [String]
      attr_accessor :service_offering_id

      # Service offering name to use for the instance
      #
      # @return [String]
      attr_accessor :service_offering_name

      # The key to be used when loging in to the vm via ssh
      #
      # @return [String]
      attr_accessor :ssh_key

      # The network_id to be used when loging in to the vm via ssh
      #
      # @return [String]
      attr_accessor :ssh_network_id

      # The network_name to be used when loging in to the vm via ssh
      #
      # @return [String]
      attr_accessor :ssh_network_name

      # The username to be used when loging in to the vm via ssh
      #
      # @return [String]
      attr_accessor :ssh_user

      # Paramters for Static NAT
      #
      # @return [String]
      attr_accessor :static_nat

      # Template uuid to use for the instance
      #
      # @return [String]
      attr_accessor :template_id

      # Template name to use for the instance
      #
      # @return [String]
      attr_accessor :template_name

      # The user data string
      #
      # @return [String]
      attr_accessor :user_data

      # The username to be used when loging in to the vm
      #
      # @return [String]
      attr_accessor :vm_password

      # The username to be used when loging in to the vm
      #
      # @return [String]
      attr_accessor :vm_user

      # Zone uuid to launch the instance into. If nil, it will
      # launch in default project.
      #
      # @return [String]
      attr_accessor :zone_id

      # Zone name to launch the instance into. If nil, it will
      # launch in default project.
      #
      # @return [String]
      attr_accessor :zone_name

      ### Firewall settings

      # comma separated list of firewall rules
      # (hash with rule parameters)
      #
      # @return [Array]
      attr_accessor :firewall_rules

      # flag to enable/disable automatic open firewall rule
      #
      # @return [Boolean]
      attr_accessor :pf_open_firewall

      ### Port forward settings

      # IP address id to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_ip_address_id

      # IP address to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_ip_address

      # public port to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_public_port

      # public port to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_public_rdp_port

      # private port to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_private_rdp_port

      # public port to use for port forwarding rule
      #
      # @return [Range]
      attr_accessor :pf_public_port_randomrange

      # private port to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_private_port

      # CIDR List string of trusted networks
      #
      # @return [String]
      attr_accessor :pf_trusted_networks

      # comma separated list of port forwarding rules
      # (hash with rule parameters)
      #
      # @return [Array]
      attr_accessor :port_forwarding_rules

      def initialize(domain_specific = false)
        # Initialize groups in bulk, re-use these groups to set defaults in bulk
        INSTANCE_VAR_DEFAULT_NIL.each do |instance_variable|
          instance_variable_set("@#{instance_variable}", UNSET_VALUE)
        end
        # Initialize groups in bulk, re-use these groups to set defaults in bulk
        INSTANCE_VAR_DEFAULT_EMPTY_ARRAY.each do |instance_variable|
          instance_variable_set("@#{instance_variable}", UNSET_VALUE)
        end

        @scheme                     = UNSET_VALUE
        @api_key                    = UNSET_VALUE
        @secret_key                 = UNSET_VALUE
        @instance_ready_timeout     = UNSET_VALUE
        @network_type               = UNSET_VALUE
        @pf_private_rdp_port        = UNSET_VALUE
        @pf_public_port_randomrange = UNSET_VALUE
        @pf_open_firewall           = UNSET_VALUE
        @expunge_on_destroy         = UNSET_VALUE

        # Internal state (prefix with __ so they aren't automatically
        # merged)
        @__compiled_domain_configs = {}
        @__finalized               = false
        @__domain_config           = {}
        @__domain_specific         = domain_specific
      end

      # Allows domain-specific overrides of any of the settings on this
      # configuration object. This allows the user to override things like
      # template and keypair name for domains. Example:
      #
      #     cosmic.domain_config "abcd-ef01-2345-6789" do |domain|
      #       domain.template_id = "1234-5678-90ab-cdef"
      #       domain.keypair_name = "company-east"
      #     end
      #
      # @param [String] domain The Domain name to configure.
      # @param [Hash] attributes Direct attributes to set on the configuration
      #   as a shortcut instead of specifying a full block.
      # @yield [config] Yields a new domain configuration.
      def domain_config(domain, attributes=nil, &block)
        # Append the block to the list of domain configs for that domain.
        # We'll evaluate these upon finalization.
        @__domain_config[domain] ||= []

        # Append a block that sets attributes if we got one
        if attributes
          attr_block = lambda do |config|
            config.set_options(attributes)
          end

          @__domain_config[domain] << attr_block
        end

        # Append a block if we got one
        @__domain_config[domain] << block if block_given?
      end

      #-------------------------------------------------------------------
      # Internal methods.
      #-------------------------------------------------------------------

      def merge(other)
        super.tap do |result|
          # Copy over the domain specific flag. "True" is retained if either
          # has it.
          new_domain_specific = other.instance_variable_get(:@__domain_specific)
          result.instance_variable_set(
              :@__domain_specific, new_domain_specific || @__domain_specific)

          # Go through all the domain configs and prepend ours onto
          # theirs.
          new_domain_config = other.instance_variable_get(:@__domain_config)
          @__domain_config.each do |key, value|
            new_domain_config[key] ||= []
            new_domain_config[key] = value + new_domain_config[key]
          end

          # Set it
          result.instance_variable_set(:@__domain_config, new_domain_config)

          # Merge in the tags
          result.tags.merge!(self.tags)
          result.tags.merge!(other.tags)
        end
      end

      def finalize!
        INSTANCE_VAR_DEFAULT_NIL.each do |instance_variable|
          # ... must be nil, since we can't default that
          instance_variable_set("@#{instance_variable}", nil) if
              instance_variable_get("@#{instance_variable}") == UNSET_VALUE
        end

        INSTANCE_VAR_DEFAULT_EMPTY_ARRAY.each do |instance_variable|
          # ... must be empty array
          instance_variable_set("@#{instance_variable}", []) if
              instance_variable_get("@#{instance_variable}") == UNSET_VALUE
        end

        # We default the scheme to whatever the user has specifid in the .fog file
        # *OR* whatever is default for the provider in the fog library
        @scheme                 = nil if @scheme == UNSET_VALUE

        # Try to get access keys from environment variables, they will
        # default to nil if the environment variables are not present
        @api_key                = ENV['COSMIC_API_KEY'] if @api_key == UNSET_VALUE
        @secret_key             = ENV['COSMIC_SECRET_KEY'] if @secret_key == UNSET_VALUE

        # Set the default timeout for waiting for an instance to be ready
        @instance_ready_timeout = 120 if @instance_ready_timeout == UNSET_VALUE

        # NetworkType is 'Advanced' by default
        @network_type           = "Advanced" if @network_type == UNSET_VALUE

        # Private rdp port defaults to 3389
        @pf_private_rdp_port     = 3389 if @pf_private_rdp_port == UNSET_VALUE

        # Public port random-range, default to rfc6335 'Dynamic Ports'; "(never assigned)"
        @pf_public_port_randomrange = {:start=>49152, :end=>65535} if @pf_public_port_randomrange == UNSET_VALUE

        # Open firewall is true by default (for backwards compatibility)
        @pf_open_firewall       = true if @pf_open_firewall == UNSET_VALUE

        # expunge on destroy is nil by default
        @expunge_on_destroy     = false if @expunge_on_destroy == UNSET_VALUE

        # Compile our domain specific configurations only within
        # NON-DOMAIN-SPECIFIC configurations.
        unless @__domain_specific
          @__domain_config.each do |domain, blocks|
            config = self.class.new(true).merge(self)

            # Execute the configuration for each block
            blocks.each { |b| b.call(config) }

            # The domain name of the configuration always equals the domain config name:
            config.domain = domain

            # Finalize the configuration
            config.finalize!

            # Store it for retrieval
            @__compiled_domain_configs[domain] = config
          end
        end

        # Mark that we finalized
        @__finalized = true
      end

      def validate(machine)
        errors = []

        if @domain
          # Get the configuration for the domain we're using and validate only that domain.
          config = get_domain_config(@domain)

          unless config.use_fog_profile
            errors << I18n.t("vagrant_cosmic.config.api_key_required") if \
               config.access_key_id.nil?
            errors << I18n.t("vagrant_cosmic.config.secret_key_required") if \
               config.secret_access_key.nil?
          end
        end

        {"Cosmic Provider" => errors}
      end

      # This gets the configuration for a specific domain. It shouldn't
      # be called by the general public and is only used internally.
      def get_domain_config(name)
        raise 'Configuration must be finalized before calling this method.' unless @__finalized

        # Return the compiled domain config
        @__compiled_domain_configs[name] || self
      end
    end
  end
end
